src/ui/public/courier/data_source/search_source.js

```diff
@@ -51,280 +51,645 @@
+            reject(error);
+            request.abort();
+          });
+
+          // Return promises created by the completion handler so that
+          // errors will bubble properly
+          return req.getCompletePromise().then(addRequest);
+        }
+
+        addRequest();
+      });
+    }
+
+    addFilterPredicate(predicate) {
+      this._filterPredicates.push(predicate);
+    }
+
+    clone() {
+      const clone = new SearchSource(this.toString());
+      // when serializing the internal state with .toString() we lose the internal classes used in the index
+      // pattern, so we have to set it again to workaround this behavior
+      clone.set('index', this.get('index'));
+      clone.inherits(this.getParent());
+      return clone;
+    }
+
+    async getSearchRequestBody() {
+      const searchRequest = await this._flatten();
+      return searchRequest.body;
+    }
+
+    /**
+     *  Called by requests of this search source when they are done
+     *  @param  {Courier.Request} request
+     *  @return {undefined}
+     */
+    requestIsStopped() {
+      this.activeFetchCount -= 1;
+    }
+
+    /**
+     * Get values from the state
+     * @param {string} name - The name of the property desired
+     * @return {any} - the value found
+     */
+    get(name) {
+      let self = this;
+      while (self) {
+        if (self._state[name] !== void 0) return self._state[name];
+        self = self.getParent();
+      }
+    }
+
+    /**
+     * Get the value from our own state, don't traverse up the chain
+     * @param {string} name - The name of the property desired
+     * @return {any} - the value found
+     */
+    getOwn(name) {
+      if (this._state[name] !== void 0) return this._state[name];
+    }
+
+    /**
+     * Change the entire state of a SearchSource
+     * @param {object|string} state - The SearchSource's new state, or a
+     *   string of the state value to set
+     */
+    set(state, val) {
+      const self = this;
+
+      if (typeof state === 'string') {
+        // the getter and setter methods check for undefined explicitly
+        // to identify getters and null to identify deletion
+        if (val === undefined) {
+          val = null;
+        }
+        self[state](val);
+      } else {
+        self._state = state;
+      }
+      return self;
+    }
+
+    /**
+     * Create a new dataSource object of the same type
+     * as this, which inherits this dataSource's properties
+     * @return {SearchSource}
+     */
+    extend() {
+      return new SearchSource().inherits(this);
+    }
+
+    /**
+     * return a simple, encodable object representing the state of the SearchSource
+     * @return {[type]} [description]
+     */
+    toJSON = function () {
+      return _.clone(this._state);
+    };
+
+    /**
+     * Create a string representation of the object
+     * @return {[type]} [description]
+     */
+    toString() {
+      return angular.toJson(this.toJSON());
```

Comment:
Couldn't we JSON.stringify here? If so, let's do that and drop one angular dependency.

Reply:
There is actually one thing different.  angular.toJson removes any variables that start with `$`.  I think that's it actually. There was a bug we ran into once because of a swap.  But I agree, we should get rid of the angular dependency.  I think there might be a native toJSON that does the same thing with the $... I'll have to look later.

But prefer not to put that in this PR.